### PR TITLE
[ROCm] clean up workspace during CI setup

### PR DESCRIPTION
**Pull request by jeffdaily**
PR #122922 added chown steps to test.sh and used the trap mechanism to ensure that, even if the test scripts fails and exits with a non-zero code, it will call the cleanup_workspace function on EXIT.

However, this doesn't work as intended when the CI job gets cancelled for eg. if a PR pushes new commits and the older commit CI job gets cancelled. The trap function doesn't get called as the test script immediately aborts.

Any subsequent jobs scheduled on the same runner then fail in the 'Checkout PyTorch' step when they try to delete the workspace.

This has been resulting in a slew of CI failures on the HUD.

This PR moves the pytorch source checkout until after the docker image has been established and adds a new step for cleaning the workspace prior to checkout.

Alternative to #123468.

cc @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang

### CODE REVIEW

File: .github/workflows/_rocm-test.yml
Comment (jithunnair-amd) [inline .github/workflows/_rocm-test.yml:97]: @jeffdaily Cleaning up at the start of the workflow seems to be a better alternative to trying to clean up at the end of the workflow, which is what PR [123468](https://github.com/pytorch/pytorch/pull/123468) was trying to do (but getting into some complications trying to make sure the cleanup is triggered reliably for every workflow run). My only concern here is that instead of removing everything inside the workspace directory, what we really need is to change the ownership permissions for the workspace directory itself to undo the effects of https://github.com/pytorch/pytorch/pull/122922/files#diff-9709e5db13aeac90c0312b2b8da34b37cf51242ec789ca959fcf1ba295a8da7aR23?

Comment (jeffdaily) [reply .github/workflows/_rocm-test.yml]: Removing workspace directory now.

Comment (jithunnair-amd) [reply .github/workflows/_rocm-test.yml]: That's better, but still poses the question: if https://github.com/pytorch/pytorch/blob/666a628bea80a4f972a2ed4c01d6a0fe66508dd4/.github/actions/checkout-pytorch/action.yml#L31 does the same thing, isn't it better to just remove `no-sudo` for the `checkout-pytorch` step, since that's what your change is doing now anyway?

Comment (jithunnair-amd) [reply .github/workflows/_rocm-test.yml]: Sorry, I realize my mistake: `checkout-pytorch` happens outside the docker container, so `sudo` won't work there.

**Review by jithunnair-amd — APPROVED** [review]

**Comment by jithunnair-amd** [comment]
@jeffdaily Getting `rm: cannot remove '/var/lib/jenkins/workspace': Device or resource busy` error in https://github.com/pytorch/pytorch/actions/runs/8618506163/job/23624510991?pr=123588#step:7:39

Looks like we need to use `chown` instead of `rm` as in PR https://github.com/pytorch/pytorch/pull/122922

**Review by atalman — APPROVED** [review]
LGTM, once all rocm test are green - please land

**Comment by atalman** [comment]
Looks like more work is needed:
```
[rocm / linux-focal-rocm6.0-py3.8 / test (default, 1, 6, linux.rocm.gpu.2, unstable)](https://hud.pytorch.org/pr/pytorch/pytorch/123588#23637672768) ([gh](https://github.com/pytorch/pytorch/actions/runs/8623145181/job/23637672768))
docker: Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: exec: "sudo chown -R 1001:1001 /var/lib/jenkins/workspace": stat sudo chown -R 1001:1001 /var/lib/jenkins/workspace: no such file or directory: unknown.
```

**Comment by malfet** [comment]
Back in CircleCI days we used to use https://hub.docker.com/_/alpine to run such ops before PyTorch or main docker images are checked out.  Perhaps that would do the trick, as one can not determine PyTorch docker hash sum before repo is checked out. 

Another suggestion: Is it possible to setup some sort of hook on the runner itself rather than keep adding those workarounds to the workflow files. 

IMO if https://github.com/actions/checkout can not be completed correctly, something is wrong with the runner state and it needs to be fixed at that point rather than trying to account for all contingencies in the workflow file

**Comment by jithunnair-amd** [comment]
> Back in CircleCI days we used to use https://hub.docker.com/_/alpine to run such ops before PyTorch or main docker images are checked out. Perhaps that would do the trick, as one can not determine PyTorch docker hash sum before repo is checked out.

@jeffdaily Hmm, this is a valid point: running "Calculate docker image" step _before_ the "Checkout PyTorch" step is likely to give unexpected results as it might use a stale pytorch copy in `/var/lib/jenkins/workspace`, or even worse, not have any pytorch copy there?
If we use an alpine image to run the `chown` command, the steps can stay in the right order.

**Comment by jeffdaily** [comment]
None of the approaches attempted by this PR were successful.  Closing in favor of #123816.

**Pull request closed without merging**
